draft-ietf-tls-external-psk-importer.md

```diff
@@ -196,17 +209,40 @@ External PSK identities are typically static by design so that endpoints may use
 lookup keying material. However, for some systems and use cases, this identity may become a
 persistent tracking identifier.
 
-# IANA Considerations
+# IANA Considerations {#IANA}
+
+This specification introduces a new registry for TLS KDF identifiers and defines the following
+KDFID values:
+
++--------------------+-------------+
+| Description        | Value       |
++--------------------+-------------+
+| Reserved           | {0x00,0x00} |
+|                    |             |
+| HKDF_SHA256        | {0x00,0x01} |
+|                    |             |
+| HKDF_SHA384        | {0x00,0x02} |
+|                    |             |
+| TLS12_PRF_MD5      | {0x00,0x03} |
+|                    |             |
+| TLS12_PRF_MD5SHA1  | {0x00,0x04} |
+|                    |             |
+| TLS12_PRF_SHA256   | {0x00,0x05} |
```

Comment:
Prior to this PR, we had exactly one more for TLS 1.2 and earlier, to reflect their existing lack of key separation. Is the thinking that we'll retroactively add in some key separation into TLS 1.2? Protocol-wise, that seems doable enough. PSKs are negotiated after and version cipher suites in TLS 1.2, the negotiation is extra straightforward. But I think we should describe this explicitly.

We also have this sentence in the introduction, which doesn't quite contradict it but may need some follow-up text:

> TLS 1.2 {{!RFC5246}}, in contrast, has no such requirement, as a PSK may be used with any hash algorithm and the TLS 1.2 PRF.

The other consideration is what we're keeping TLS 1.2 compatibility around for. These are the scenarios I can imagine:

1. This is a new application with no legacy TLS 1.2 PSK deployment at all. It can just assume TLS 1.3+.
2. This is a new application with no legacy TLS 1.2 PSK deployment, but for some odd reason it can't just assume TLS 1.3+?
3. This is an application which previously used TLS 1.2 PSKs. It wishes to smooth migrate to TLS 1.3, which means dual-1.2/1.3 endpoints must exist for a period of time.
4. Same as above, but this time the underlying TLS library wishes to enable TLS 1.3 by default, so no API calls to the TLS stack may change.

Case 1 is easy. It just doesn't care about TLS 1.2 logic in this draft, either before or after this PR.

Case 2 is odd. It can use the TLS 1.2 logic in this draft. Before this PR is definitely fine as it requires no new TLS 1.2 behavior. After this PR may be tricky, depending on exactly why you can't assume TLS 1.3. If it's because, say, you need to run on an older version of some OS without TLS 1.3, you're limited by what the older TLS 1.2 API exposed. That may or may not be sufficient for this scheme. (E.g. OpenSSL does some callback thing, so you might be able to implement it?? Not sure. But declaratively configuring a set of PSKs won't know to match the KDF.) If you need TLS 1.2 but are able to make code changes to your TLS 1.2 implementation, I suppose after this PR is fine, though I'm confused why you don't use your code change powers to get TLS 1.3.

Case 3 also isn't helped by this draft, before or after the PR. The existing TLS 1.2 PSK endpoints don't know how to do the IPSK derivation, so, one way or another, the dual-1.2/1.3 endpoints need to speak the old-style derivation at TLS 1.2, not the new style. Depending on how exactly the PSKs are configured, one strategy is to simply configure two separate PSKs, one legacy for TLS 1.2 and one new for TLS 1.3+. If that's not an option and you truly only get one root secret, you are sad. You could pass the same root secret into the legacy and modern path, at the cost of some cross-protocol risk. That's the problem I was trying to address [here](https://tools.ietf.org/html/draft-davidben-tls-universal-psk-00#section-3).

Case 4 is a harder version of case 3. The library's options are either to say, as a compatibility hack, disable TLS 1.3 for applications that called the `SetTLS12PreSharedKey()` function, or you've got the "one root secret" problems above.

Reply:
I agree with your points above. It seems to me that the most realistic case is (1), and, at that point, you might as well just prohibit importing for TLS 1.2 and prior versions. That lets us drop a lot of cruft. (I applied that change in the latest update!)